### ISSUE TITLE
Provide abstraction/interface to implement resource isolation for custom resources

### ISSUE THREAD

**Issue by JasonWayne** [ray-project/ray#7204]
Currently, ray can use custom resources for new kind of hardwares like TPU. But to my best knowledge, custom resoures can not provide real resource isolation in ray like GPU in the following code:

```python
# ray/_raylet.pyx
cdef execute_task(
        CTaskType task_type,
        const CRayFunction &ray_function,
        const unordered_map[c_string, double] &c_resources,
        const c_vector[shared_ptr[CRayObject]] &c_args,
        const c_vector[CObjectID] &c_arg_reference_ids,
        const c_vector[CObjectID] &c_return_ids,
        c_vector[shared_ptr[CRayObject]] *returns):
     ...
     # Automatically restrict the GPUs available to this task.
     ray.utils.set_cuda_visible_devices(ray.get_gpu_ids())
     ...
```

Can we provide some interface for developers to implement resource isolation for custom resources?

**Comment by JasonWayne** [comment]
@ericl @richardliaw @robertnishihara Pls help to check this.

**Comment by ericl** [comment]
I think a more general way would be to have a `pre_execute_hook` that can be globally registered for all tasks / actors. Then you can put whatever you want in there including resource isolation settings.

**Comment by richardliaw** [comment]
Yeah, I actually think that's a great idea... we have makeshift versions in rllib and raysgd already.

**Comment by richardliaw** [comment]
Though the thing, for say CUDA devices, how do you know what GPU are you given?

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
